src/databricks/labs/ucx/progress/install.py

```diff
@@ -0,0 +1,81 @@
+import datetime as dt
+import logging
+from dataclasses import dataclass
+
+from databricks.labs.lsql.backends import Dataclass, SqlBackend
+from databricks.sdk.errors import InternalError
+from databricks.sdk.retries import retried
+
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, kw_only=True)
```

Comment:
let's separate job runs and records

```python
@dataclass(frozen=True, kw_only=True)
class WorkflowRun:
    started_at: dt.datetime
    finished_at: dt.datetime
    workspace_id: int
    workflow_name: str # migration-progress, assessment, etc, etc
    workflow_id: int
    workflow_run_id: int # <=== this column to be used for joining with historical records, create a view to pre-join these. i think you called it "snapshot" or something.
    run_as: str
    status: str
```

essentially, it'll be

```
WITH level0 AS (
  SELECT workspace_id, workflow_id, MAX(finished_at) AS finished_at FROM workflow_runs GROUP BY 1,2
), level1 AS (
  SELECT * FROM level0 LEFT JOIN workflow_runs USING (workspace_id, workflow_id, finished_at)
)
SELECT * FROM level1
```

Reply:
That is a separate issue: https://github.com/databrickslabs/ucx/issues/2600

Reply:
then we need that one land `main` first, as history needs `workflow_run_id` to join on. other PR struggles with `started_at` and `finished_at` fields. same for `run_as`, actually

Reply:
https://github.com/databrickslabs/ucx/pull/2754